File: c/chapitre_5/exo52.c
/*
Écrire un programme qui crée un thread qui alloue un tableau d’entiers,
initialise les éléments par des entiers aléatoires entre 0 et 99, et retourne le tableau d’entiers.
*/

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <time.h>

#define TAILLE 4

void *ma_fonction_thread(void * arg) ;

int main()
{
    pthread_t mythread ;
    int *tableau = NULL ;
    srand(time(NULL));

    //Lancement du thread 
    pthread_create(&mythread , NULL , ma_fonction_thread , (void*)tableau ) ;
    pthread_join(mythread , (void*)&tableau ) ;

    // test afficahage du tableau 
    for (int i = 0; i < TAILLE ; i++)
    {
        printf("[%d] ",tableau[i] ) ;
    }
    
    return (0);
}

void *ma_fonction_thread(void *arg)
{
    int *tab = (int*)arg ;

    tab = malloc(TAILLE * sizeof(int));

    for (int i = 0 ; i < TAILLE ; i++)
    {
        tab[i] = rand() % 100 ;
    }
    
    pthread_exit((void*)tab) ;
}

File: python/chapitre_5/exo52.py
"""
Écrire un programme qui crée un thread qui alloue un tableau d'entiers,
initialise les éléments par des entiers aléatoires entre 0 et 99, et retourne le tableau d'entiers.
"""

import threading
import random
import time

TAILLE = 4

def ma_fonction_thread(arg, result):
    tab = [random.randint(0, 99) for _ in range(TAILLE)]
    result.extend(tab)
    # pthread_exit equivalent : return


def main():
    random.seed()
    tableau = []

    # Lancement du thread
    thread = threading.Thread(target=ma_fonction_thread, args=(None, tableau))
    thread.start()
    thread.join()

    # test affichage du tableau
    for i in range(TAILLE):
        print(f"[{tableau[i]}] ", end='')

    return 0

main()